File: 15.3sum/solution.cpp
class Solution {
public:
	vector<vector<int>> threeSum(vector<int>& nums) {
		vector<vector<int>> ans2;
		if (nums.empty())
			return ans2;
		int *p = &*nums.begin();
		sort(p, p + nums.size());
		int len = nums.size();
		set<vector<int>> ans;
		for (int i = 0; i < len; i++)
		{
			int j = i + 1;
			int k = len - 1;
			while (j < k)
			{
				int result = nums[i] + nums[j] + nums[k];
				if (result == 0)
				{
					int arr[] = {nums[i], nums[j], nums[k]};
					ans.insert(vector<int>(arr, arr + 3));
					j++;
					k--;
				} else if (result > 0)
					k--;
				else
					j++;
			}
		}
		for (set<vector<int>>::iterator it = ans.begin(); it != ans.end(); it++)
		{
			ans2.push_back(std::move(*it));
		}
		return ans2;
	}
};

File: 15.3sum/solution.go
func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	m := make(map[int][]int)
	for i, n := range nums {
		m[n] = append(m[n], i)
	}
	var result [][]int
	for i, n := range nums {
		if i > 0 && n == nums[i - 1] {
			continue
		}
		for j := i + 1; j < len(nums); j++ {
			if j > i + 1 && nums[j] == nums[j - 1] {
				continue
			}
			if ks,ok := m[-n-nums[j]]; ok {
				for _, k := range ks {
					if k > j {
						result = append(result, []int{n, nums[j], -n-nums[j]})
						break
					}
				}
			}
		}
	}
	return result
}
